resolver.go

```diff
@@ -221,28 +235,58 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 		return
 	}
 
+	proto := w.LocalAddr().Network()
 	if resp == nil {
-		if len(r.extDNS) == 0 {
-			return
-		}
-
-		num := maxExtDNS
-		if len(r.extDNS) < maxExtDNS {
-			num = len(r.extDNS)
-		}
-		for i := 0; i < num; i++ {
-			log.Debugf("Querying ext dns %s:%s for %s[%d]", w.LocalAddr().Network(), r.extDNS[i], name, query.Question[0].Qtype)
+		for i := 0; i < maxExtDNS; i++ {
+			extDNS := &r.extDNSList[i]
+			if extDNS.ipStr == "" {
+				break
+			}
+			log.Debugf("Querying ext dns %s:%s for %s[%d]", proto, extDNS.ipStr, name, query.Question[0].Qtype)
+
+			extConn = extDNS.extConn
+			if extConn == nil || proto == "tcp" {
+				r.sb.execFunc(func() {
+					addr := fmt.Sprintf("%s:%d", extDNS.ipStr, 53)
+					extConn, err = net.DialTimeout(proto, addr, extIOTimeout)
+				})
+				if err != nil {
+					log.Debugf("Connect failed, %s", err)
+					continue
+				}
+				// For udp clients, persist the connection so that it can be reused
+				// for further queries
+				if proto == "udp" {
+					extDNS.extConn = extConn
```

Comment:
concurrent calls to `ServeDNS` is possible and the possibility for race is in `extConn = extDNS.extConn`.  But the DNS queries from the container (even if they originate in parallel) are still serialized over the same socket.. So I think its hard to hit. Even if we assume its possible, the worst case scenario is one extra UDP connection for that container. I am not sure if we need locking primitives here.